sound/soc/sof/control.c

```diff
@@ -195,3 +210,89 @@ int snd_sof_bytes_put(struct snd_kcontrol *kcontrol,
 	pm_runtime_put_autosuspend(sdev->dev);
 	return 0;
 }
+
+/* IO handlers for switch kcontrol handlers */
+int snd_sof_switch_get(struct snd_kcontrol *kcontrol,
+		       struct snd_ctl_elem_value *ucontrol)
+{
+	struct soc_mixer_control *sm =
+		(struct soc_mixer_control *)kcontrol->private_value;
+	struct snd_sof_control *scontrol = sm->dobj.private;
+	struct snd_sof_dev *sdev = scontrol->sdev;
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	unsigned int i, channels = scontrol->num_channels;
+
+	pm_runtime_get_sync(sdev->dev);
+
+	/* get all the mixer data from DSP */
+	snd_sof_ipc_get_comp_data(sdev->ipc, scontrol, SOF_IPC_COMP_GET_VALUE,
```

Comment:
@plbossart I am not sure I follow the question. The switch_get() function is just like the volume_get() function to read the values from the DSP. Why do you think it is not necessary to read the switch status values from the DSP? Could you please elaborate?

Reply:
the premise for kcontrols is that we only do an IPC if strictly necessary, cause IPC can be slow. This is why we have regmap and caches.
So we should only get the values from the DSP if they are volatile and can change on their own in the firmware. If the driver already has the information, it can provide it back directly.
If you look at the put() case it's exactly what you do, you only send an IPC when the value changed, so why do we need an IPC for get()?

Reply:
@plbossart Thanks for the explanation. The put() case was a bit more straightforward to make sure that we send the IPC only when the values are changed. 

But I get what you mean now. The switch status is never changed in the firmware on its own. So it makes sense to just return the value in the driver. I'll update it.